### DISCUSSION TITLE
Daisy chain

### DISCUSSION THREAD

**Discussion by coderatchance** [SpenceKonde/megaTinyCore#1146] [Future Development Discussion]
Hello guys, 

I would like to implement a daisy chain, namely transfering a function via the ambient light sensor,
There are two tinys running the same prox sensor and ALS sensor;

When the prox goes above on the 1st, the led turns on and this should fire the 2nd tiny's ALS sensor and turn on the led.
The led on the 1st tiny fades away, while the led on the second tiny stays on!

Any ideas how to tackles this ? Some ideas are to go with the flip.flop, but not sure how to get around it?

best.

**Comment by MX682X** [comment]
A couple questions:
1. What problem are you trying to solve with the MCUs?
2. Why are you using two seperate MCUs?
3. Which Tinies are you using/planning to use?
4. When does the LED on the second Tiny turn off again?

**Reply by coderatchance** [reply]
Hi MX682X!

Thanks for your consideration of my problem.

The main idea is to have the master and slave tiny, which are 3206. Two individual breadboards with sensor for als and prox.
The mcu actually controls the led via open fet, open collector, and follows a simple following procedure:

if master turns on the led (manually via prox), after few ms turns it off automatically.
During that 100ms the slave detect the light via the als and takes over the led, it turn it on, and waits for another manual off, via the prox. 

I am using two tinys because I need two seperate, individuals tinys at different location...

Does this help? If not, please let me know...

best.

**Reply by MX682X** [reply]
I can't find a 3206, I assume it's a 3216, as a 1606 would require more typos. Assuming both sensors output analog values, I would suggest to use AC0 and AC1 for the ALS and PROX respectivly. 
Select AC0 as Input for the LUT0, and AC0 as input for LUT1, with an Edge-Detector and Filter. 
Set the SEQCTRL0 to JK Flip-Flop. 
Enable LUT0-OUT (PA4).

In the best case scenario, you can put the whole MCU into Stand-by sleep mode and let the periphery do all the logic. (Main Clock would have to bet set to the 32kHz ULP oscillator for the filters).

However, this solution requires you to choose the passive components carefully. You can supply the threshold voltage through a resistor divider, or use the VREF of the microcontroller (I recommend 1.1V, as the absolute accuracy deviation is smaller) (which won't work if you are battery powered, then you need a resitor bridge)

The above was for the slave. The master would require a Timer, e.g. TCB instead AC1 to Kill the Flip-Flop Output.

Should it be 1606, thigs will be trickier, as it has only one AC...

**Reply by coderatchance** [reply]
Thank you for a good idea.
 I will have to go with what I have, which is (sorry for the typo) at tiny 3216 (internal osc).
I have the protoboard and tiny with the sensor (vcnl4040 package) ready and it works well via the i2c... so I guess the ACs are not necessary to get a working solution... 

Anyway, I am not really sure how to tackle the Flip.Flop which is actually what I need, you are right indeed, any ideas?

Best.

**Reply by MX682X** [reply]
I see. Is your project battery powerd, as in needs a low power consumption? If yes, is the Interrupt pin from the sensor connected to the MCU?

**Reply by coderatchance** [reply]
No, it is power supply  12V -> 5V , but there is an interrupt between the sensor and the mcu!

PS I am in arduino ide

**Reply by MX682X** [reply]
Well, you can simply use the interrupt to time the sensor read, and millis() to time the signal. A simple digitalWrite(LED_PIN) will do the trick of a "Flip-Flop"

**Comment by coderatchance** [comment]
Let me recap; so if the sensor PROX goes above threshold, fire the interrupt and start summing the time; current + start time. Once the PROC  below the threshold, fire or clear the interrupt? That should give me the time for the proximity?The ALS is done as a toggle so far. Not sure how to advance from here and tackle the flip-flop?Anyway, it sounds very optimistic for me.Can you throw an example, I am not good with the interrupts; always lost when it comes to clearing them especially in the Arduino, etc.

**Reply by MX682X** [reply]
here is a little pseudocode for master:

pinmode(INT_PIN, INPUT_PULLUP);

if (digitalRead(INT_PIN) == 0) {
  // read sensor
  if (proxy > proxy_threshold) {
    stop_time = millis() +100;
    digitalWrite(LED_PIN, HIGH);
  }
}
if (millis() > stop_time) {
  dgitalWrite(LED_PIN, LOW);
}

here is a little pseudocode for slave:

pinmode(INT_PIN, INPUT_PULLUP);

if (digitalRead(INT_PIN) == 0) {
  // read sensor
  if (proxy > proxy_threshold) {
    digitalWrite(LED_PIN, LOW);
  } else if (light > light_threshold) {
    digitalWrite(LED_PIN, HIGH);
}

**Comment by coderatchance** [comment]
Thanks, will adopt it and try to use it in my hw setup. Need to 1st wire up the interrupt pin, etc… and code it in the setup along with the sensor readings, etc. Will report back… There is a code for master/slave, although it is rather clear what each is doing, not sure how will they run together…Best!

**Reply by coderatchance** [reply]
With this code I can get to the stage where master turns on , after few msec it turns off, while the slave, picks up when the led was on, and turns on its own led, and waits for the wave to turn led off... however, it does not work vice versa in one code...

**Comment by coderatchance** [comment]
Hello @MX682X

I have deleted the previous code, created new, however, it does not work, I am returning here for help, somehow I cannot get my head around this. Can you help?

```
// LOOP3 LOOP3 LOOP3 LOOP3 LOOP3 LOOP3 LOOP3 LOOP3 LOOP3 LOOP3 LOOP3 LOOP3 LOOP3 LOOP3 LOOP3 LOOP3 LOOP3 LOOP3 LOOP3 LOOP3 LOOP3 LOOP3
void loop() {

  Serial.print(proxState);
  Serial.println(alsState);

 // timer to turn off the led?
unsigned long msec = millis ();
  if (tmr && msec - msec0 >= MsecPeriod && ledSet(On)) {
 Serial.println ("timer expired");
 tmr = 0;
Serial.print("led is off");
ledState = ledSet (Off);
alsState = 0;

}

  if (ALS_HT < Sensor.getAmbient() && !proxState && !alsState) {
    alsState = 1;
    Serial.println("the ALS is on ");
    //ledSet(On);
  }

  if ( alsState and (proxT < Sensor.getProximity())  ) {
    proxState = 1;
    if (proxState && alsState) {
      Serial.println("led off");
      proxState = 0;
      alsState = 0;
    }
  }


  // -----------------
  // turn on the led by a wave
  // after some time, turn off the led
  // get of the way

  if (proxT < Sensor.getProximity() && !alsState)  {
    if (! proxState)  {                // was off
      if (! tmr)  {
        tmr   = true;                 // enable timer
        msec0 = msec;
        Serial.println(" led is on");
        //ledSet (On);
      }
      else  {
        tmr = false;                 // disable timer
        Serial.print(" led is off due to tmr ");
        //ledSet (Off);
      }
    }
    proxState = true;
  }
  else
    proxState = false;

  delay(500);

} // end of  LOOP3
```

**Reply by MX682X** [reply]
Sorry for the delay, had a busy week.

I have no experience with the sensor you are using, but If I understand correctly, you want the LED of the slave to turn off when the proximity sensor is below threshold. If you are fine with a little delay after removing the hand, this should work. Of course I can't check for compilation errors, as I don't have the complete code. Should this not work, I highly recommend you to use a Finite State Machine instead of boolean checks.

```
// LOOP3 LOOP3 LOOP3 LOOP3 LOOP3 LOOP3 LOOP3 LOOP3 LOOP3 LOOP3 LOOP3 LOOP3 LOOP3 LOOP3 LOOP3 LOOP3
void loop() {
  // timer to turn off the led?
  unsigned long msec = millis();
  if (msec > msec0) {
    Serial.println("timer expired");
    Serial.print("led is off");
    msec0 = (uint32_t)-1; // avoid refiring
    ledSet(Off);
  }

  if (ALS_HT < Sensor.getAmbient()) {
    Serial.println("the ALS is on ");
    ledSet(On);
  }

  // -----------------
  // turn on the led by a wave
  // after some time, turn off the led
  // get of the way

  if (proxT < Sensor.getProximity()) {
    msec0 = msec + MsecPeriod - 1;
    ledSet (On);
  }

  delay(100);

}  // end of  LOOP3
```

**Reply by coderatchance** [reply]
Hello!

Have the solution and works well, see below, I would like to add the option of not only turning the led off at the other switch, which is as it is done now, but also at first one. For example, I turn on the led on the first switch, and I am on the way to the other, to turn it off, but I have decided to go back and turn off the led on the first one, Any ideas?

Best.

```
  //timer to turn of the proxState
  if ( proxState && (msec1 - msec0 >= 500 ) ) {
    Serial.println(" timer passed ");
    proxState = false;
    alsState = false;
    msec0 = msec1;
  }


  // ALS reading
  // presumably the led is off
  if (!lock && !proxState && ALS_HT < Sensor.getAmbient()) {  // lock

    //wait two seconds and check if the light is still present?
    if ( (msec - msec0 >= 1000) && getAmbient() ) {
      msec0 = msec;
      //lock = false;
      Serial.print("setting the als State  ");
      //then set the alsState ?
      alsState = true;
    }
  }

  // singl wave with timing above proxT for 500ms, toggles proxState!
  if (proxT < Sensor.getProximity() && (msec - msec0 >= 400) ) {
    msec0 = msec;
    //lock = true;

    if (proxState) {
      proxState = false;
    }
    else if (!proxState) {
      proxState = true;
    }
  }

  //*************turning on the led old ledAction....
  // 11
  if (proxState && alsState) {
    ledSet(Off);
    //proxState = 0;
    alsState = 0;
  }
  // 10
  if (proxState && !alsState) {
    ledSet(On);
  }
  // 01
  if (!proxState && alsState) {
    ledSet(On);
  }
  // 00
  if (!proxState && !alsState) {
    ledSet(Off);
    lock = true;
  }
  //**************turning on the led

  if (Sensor.getAmbient() < ALS_HT) {
    lock = false;
    //lock open
  }

} // end of  LOOP3
```

**Reply by MX682X** [reply]
>  Any ideas?

use a state-machine with an enum together with an if/if else or switch/case. This will make it easier to handle inputs and significatly easier to expand compared to your current approach.

**Reply by coderatchance** [reply]
Ok, I guess you are referring to this part?

```
  //*************turning on the led old ledAction....
  // 11
  if (proxState && alsState) {
    ledSet(Off);
    //proxState = 0;
    alsState = 0;
  }
  // 10
  if (proxState && !alsState) {
    ledSet(On);
  }
  // 01
  if (!proxState && alsState) {
    ledSet(On);
  }
  // 00
  if (!proxState && !alsState) {
    ledSet(Off);
    lock = true;
  }
  //**************turning on the led
```

**Reply by MX682X** [reply]
Basically, yes.
